### PR TITLE
Sienna Kopf Make It Work Independent challenge 1

### PR DESCRIPTION
**Pull request by sienna-kopf**
I think I did well in this challenge. I got through all the required stories and both extensions. 

I was a little unsure of where to house some functionality for this challenge, especially the location and name of action for adding a contestant to a project so I would love some clarification there.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,3 +1,8 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  get "/projects/:project_id", to: 'projects#show'
+  get "/contestants", to: 'contestants#index'
+
+  post "/projects/:project_id/contestants", to: 'contestants#add'
```
Comment (BrianZanti) [inline config/routes.rb:7]: When you are adding a contestant to a project, what you are really doing is creating a new `contestant_project`, so you could have this route be `post "/projects/:project_id/contestant_projects", to: 'contestant_projects#create'`. When I see a Post route where the path ends in `contestants`, based on my restful principles I would guess that route creates a contestant which it doesn't

File: spec/features/contestants/index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    ContestantProject.create(contestant_id: @gretchen.id, project_id: @upholstery_tux.id)
+    ContestantProject.create(contestant_id: @kentaro.id, project_id: @upholstery_tux.id)
+    ContestantProject.create(contestant_id: @kentaro.id, project_id: @boardfit.id)
+    ContestantProject.create(contestant_id: @erin.id, project_id: @boardfit.id)
+  end
+
+  it 'can display contestant name and names of the projects they have worked on' do
+    visit "/contestants"
+
+    expect(page).to have_content(@jay.name)
```
Comment (BrianZanti) [inline spec/features/contestants/index_spec.rb:31]: This is a small thing, but I would put each of these expectations for the names inside the corresponding `within` block.

File: app/models/project.rb [hunk 3]
```diff
@@ -1,4 +1,18 @@
...
   belongs_to :challenge
+
+  has_many :contestant_projects
+  has_many :contestants, through: :contestant_projects
+
+  def contestant_experience
+    total_experience = contestants.sum do |contestant|
+      contestant.years_of_experience
+    end.to_f
+    total_experience / contestants.count.to_f
```
Comment (BrianZanti) [inline app/models/project.rb:12]: Try to avoid using Ruby to do calculations. Making calculations is far more efficient if you let the database handle it:

```
contestants.average(:years_of_experience)
```

This code will generate a SQL statement to calculate the average for you. It uses the ActiveRecord#calculate method rather than the Ruby Array#sum method. We will talk more about this when we get to grouping and aggregating.

File: app/models/project.rb [hunk 4]
```diff
@@ -1,4 +1,18 @@
...
+  end
+
+  def add_contestant(contestant_id)
```
Comment (BrianZanti) [inline app/models/project.rb:15]: This is great! You are making an abstraction for how contestants can be added to projects.

File: app/models/contestant.rb [hunk 5]
```diff
@@ -1,3 +1,12 @@
 class Contestant <ApplicationRecord
   validates_presence_of :name, :age, :hometown, :years_of_experience
+
+  has_many :contestant_projects
+  has_many :projects, through: :contestant_projects
+
+  def projects_list
```
Comment (BrianZanti) [inline app/models/contestant.rb:7]: Great job adding this method! This makes your work in the view much easier 👍

File: app/controllers/contestants_controller.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+class ContestantsController < ApplicationController
+  def index
+    @contestants = Contestant.all
+  end
+
+  def add
+    project = Project.find(params[:project_id])
+    if project.contestants.include?(Contestant.find(params[:contestant_id]))
```
Comment (BrianZanti) [inline app/controllers/contestants_controller.rb:8]: Good job handling the edge case when a contestant is already on a project. Ideally, I would also want to see a test for this as well, but understandable that you don't have one since this is a timed challenge. Make sure you are doing it in the project!

**Review by BrianZanti — COMMENTED** [review]
Great job Sienna! The code looks very well tested and fully functional. I would encourage you to keep writing those model methods that push work onto the models and make your controllers and views nice and clean.

File: app/views/projects/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+<h1> <%= @project.name %> </h1>
+<p>Material: <%= @project.material %></p>
+<p>Challenge Theme: <%= @project.challenge.theme %></p>
+<p>Number of Contestants: <%= @project.contestants.count %></p>
```
Comment (BrianZanti) [inline app/views/projects/show.html.erb:4]: This is another opportunity to add a model method. This is technically database logic that should live in the model.
